lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -2759,6 +2790,62 @@ bool HandshakeBuilder::visitHandshake(handshake::LoadOp op) {
   return true;
 }
 
+/// Please refer to test_pack_unpack.mlir test case.
+bool HandshakeBuilder::visitHandshake(PackOp op) {
+  ValueVector tuple = portList.back();
+  Value tupleData = tuple[2];
+
+  auto bundleType = tupleData.getType().dyn_cast<BundleType>();
+
+  // Create subfields for each bundle element
+  ValueVector elements;
+  for (size_t i = 0, e = bundleType.getNumElements(); i < e; ++i) {
+    elements.push_back(rewriter.create<SubfieldOp>(insertLoc, tupleData, i));
+  }
+
+  // Collect all input ports.
+  SmallVector<ValueVector *, 4> inputs;
+  for (unsigned i = 0, e = portList.size() - 1; i < e; ++i)
+    inputs.push_back(&portList[i]);
+
+  // Connect each input to the corresponding part of the output bundle
+  for (auto [element, input] : llvm::zip(elements, inputs)) {
+    rewriter.create<ConnectOp>(insertLoc, element, (*input)[2]);
+  }
+
+  return buildJoinLogic(inputs, &tuple);
+}
+
+/// Please refer to test_pack_unpack.mlir test case.
+bool HandshakeBuilder::visitHandshake(UnpackOp op) {
+  ValueVector tuple = portList[0];
+  Value tupleData = tuple[2];
+  unsigned portNum = portList.size();
+
+  auto bundleType = tupleData.getType().dyn_cast<BundleType>();
+
+  // Create subfields for each bundle element
+  ValueVector elements;
+  for (size_t i = 0, e = bundleType.getNumElements(); i < e; ++i) {
+    elements.push_back(rewriter.create<SubfieldOp>(insertLoc, tupleData, i));
+  }
+
+  // Collect all output ports.
+  SmallVector<ValueVector *, 4> outputs;
+  for (int i = 1, e = portNum - 2; i < e; ++i)
+    outputs.push_back(&portList[i]);
+
+  // Connect each bundle element to the corresponding output
+  for (auto [element, output] : llvm::zip(elements, outputs)) {
+    rewriter.create<ConnectOp>(insertLoc, (*output)[2], element);
+  }
+
+  auto clock = portList[portNum - 2][0];
+  auto reset = portList[portNum - 1][0];
```

Comment:
Side note - ughhh, can't wait until the `handshake-to-hw` rewrite gets up to speed, so assumptions like this can be reworked.

Reply:
+1
This is especially weird because it depends on a trait/interface added somewhere else. If someone removes `HasClock` everything explodes without any reasonable warnings :S